### ISSUE TITLE
Inconsistent use of HttpClientException and returning refused payment

### ISSUE THREAD

**Issue by johanssontobias** [Adyen/adyen-dotnet-api-library#550]
Is there a reason why some refusal reasons are returned as a refused, however, valid payment response while others are wrapped in a HttpClientException?

* If I make a payment request with a card with unsuffient funds, I get a payment response with a refusal reason stating the payment was refused. 
* If I make a payment request with a payment method I do not have configured on the merchant account, I get a HttpClientException thrown with Code = 500 wrapping the payment response.
* If I make a payment request with an unidentifiable card number, I get a HttpClientException thrown with Code = 422 wrapping the payment response.

I feel this is rather inconsitent and unnecessarily complex. The payment exists (however refused) on Adyen's end and if I want to have a 1-1 relationship on our end (i. e. saving all payments in our DB that exists on Adyen's end), I need to create a wrapper to try-catch this behavior.

```csharp
public interface IAdyenCheckout
{
    Task<PaymentResponse> PaymentsAsync(
    	PaymentRequest request, 
    	CancellationToken cancellationToken = default);
}

public class AdyenCheckout : IAdyenCheckout
{
    private readonly Checkout _checkout;

    public AdyenCheckout(Client client)
    {
        _checkout = new Checkout(client);
    }

    public async Task<PaymentResponse> PaymentsAsync(
    	PaymentRequest request, 
    	CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();

        try
        {
            return await _checkout.PaymentsAsync(request);
        }
        catch (HttpClientException exception)
        {
            var responseBody = JsonConvert.DeserializeObject<AdyenErrorResponseBody>(exception.ResponseBody);

            if (responseBody.ErrorCode.StartsWith("905"))
            {
                return new PaymentResponse(
                    amount: request.Amount,
                    merchantReference: request.Reference,
                    pspReference: responseBody.PspReference,
                    refusalReason: responseBody.Message, // Payment Details are not supported
                    resultCode: PaymentResponse.ResultCodeEnum.Error);
            }

            if (responseBody.ErrorCode.StartsWith("101"))
            {
                return new PaymentResponse(
                    amount: request.Amount,
                    merchantReference: request.Reference,
                    pspReference: responseBody.PspReference,
                    refusalReason: responseBody.Message, // "Invalid card number"
                    resultCode: PaymentResponse.ResultCodeEnum.Refused);
            }

            throw;
        }
    }
}
```

For the payment examples above, I'd have simply expected a payment response with RefusalReason =  "Invalid card number" or "Payment Details are not supported" and not an exception.

/ Tobias

**Comment by adbre** [comment]
I am troubleshooting similar issues and I have noticed that the real client implementation `HttpUrlConnectionClient` is throwing an HttpClientException, while the unit tests in this repository for Checkout is not expecting an exception to be thrown 
https://github.com/Adyen/adyen-dotnet-api-library/blob/76fe7e64883dba049fca9eb154e7486c3e36beca/Adyen.Test/CheckoutTest.cs#L181

The mock client used for the unit tests only loads a response and returns it, and does not check for success status code, which the (obsolete) HttpWebRequest does and throws a WebException, which Adyens HttpUrlConnectionClient catches and wraps in a HttpClientException.

In summary; to me it appears the _intended_ behaviour, based on the unit test behaviours, in the Adyen .NET SDK is to _not_ for the client to throw a HttpClientException.

Not tested it myself yet, but I think a custom implementation of IClient wich uses the recommended HttpClient instead of legacy HttpWebRequest would, at least partially, solve the issue and produce a more consistent behaviour.

**Comment by johanssontobias** [comment]
There is an open PR to use HttpClient instead of the obsoleteHttpWebRequest. https://github.com/Adyen/adyen-dotnet-api-library/pull/477. The PR has been open for quite a while though...

**Comment by johanssontobias** [comment]
@AlexandrosMor @martinsrenato @rikterbeek @zaiddreakh @wboereboom 

Any comments on this?

**Comment by AlexandrosMor** [comment]
Hello @adbre 

We are aware that Microsoft recommends to not use HttpWebRequest for new development but instead, use the [System.Net.Http.HttpClient](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=net-6.0) class. We are going to investigate again the possibility to replace it.

Indeed the mock client used for the unit tests only loads a response and returns it, and does that because at the moment that we are implementing the tests git hub have not introduced the github secrets so we could add integration tests which we run them internal. 


Indeed the

**Comment by AlexandrosMor** [comment]
Hello @johanssontobias,

The scenarios that you described are api payment responses? In the example that you provide the client in the exception handling creates new payment responses and the library should not do that. The library is only passing the request and the response but not interfere with the data

kind regards,
Alexandros
Adyen
